Move customer validation into github-event

martindekov opened the pull request:
Moving customer validation out of github-push
and into github-event depending on commit #271

Signed-off-by: Martin Dekov (VMware) <mdekov@vmware.com>

## Description

This commit extends #271 and if merged closes #163. Currently the build fails since the struct is not vendored.

> Note: Not sure if the `Customers` struct in #271 would be used elsewhere. Should I add it into the handler directly? 

Also added additional check for `\r` since if making changes to a file through github UI it will append `\r\n` to the end of the line rather than only `\n` as expected. I personally have separate repository that contains my customers for faster updating.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

Have the change in image `martindekov/github-event:163` and deployed image on my instance.

Added my own Customers file and tested with case sensitive name existing in the file
```
Push - {refs/heads/master {exmpl MartinDekovWat/exmpl https://github.com/MartinDekovWat/exmpl.git false {MartinDekovWat 42210992+MartinDekovWat@users.noreply.github.com}} d86e067fc83d4ada19ce05cdf5edc141d35d0076 {312235}}, git-tar status: 202
```
And my account non existing in file
```
Customer: martindekov not found in CUSTOMERS file via https://raw.githubusercontent.com/martindekov/mycustomers/master/CUSTOMERS
```

## How are existing users impacted? What migration steps/scripts do we need?

N.A. moving logic out of the `github-push`

## Checklist:

I have:

- [ ] updated the documentation and/or roadmap (if required)
- [x] read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] signed-off my commits with `git commit -s`
- [ ] added unit tests

ivanayov commented:
The build fails with `function/handler.go:51:15: undefined: sdk.Customer`
You need to take #271

alexellis commented:
@martindekov why does this fail the CI?

martindekov commented:
Because this extends #271 once merged I will vendor the SDK and it wont fail.

martindekov commented:
Ah ok I will

ivanayov commented:
Looks good

alexellis commented:
Looks ready to merge.

alexellis commented on github-event/handler.go:
```diff
@@ -238,3 +268,56 @@ func forward(req []byte, function string, headers map[string]string) (string, in
+
+// getCustomers reads a list of customers separated by new lines
+// who are valid users of OpenFaaS cloud
+func getCustomers(customerURL string) ([]string, error) {
```
How is this different from what we do for Derek? Can you take a look and compare?

bartsmykla commented on github-event/handler.go:
```diff
@@ -37,6 +38,35 @@ func Handle(req []byte) string {
+		}
+
+		customer := sdk.Customer{}
+		json.Unmarshal(req, &customer)
```
There is missing error handling

bartsmykla commented on github-event/handler.go:
```diff
@@ -37,6 +38,35 @@ func Handle(req []byte) string {
+
+		found = validCustomer(customers, customer.Sender.Login)
+
+		if !found {
```
Instead of `if !found {` you can do:
```
if !validCustomer(customers, customer.Sender.Login) {
```

martindekov replied:
Good idea!

bartsmykla commented on github-event/handler.go:
```diff
@@ -238,3 +268,56 @@ func forward(req []byte, function string, headers map[string]string) (string, in
+	}
+
+	return found
+}
```
I would write it:
```
func validCustomer(customers []string, owner string) bool {
	for _, customer := range customers {
		if len(customer) > 0 && strings.EqualFold(customer, owner) {
			return true
		}
	}

 	return false
}
```

martindekov replied:
Done ^

bartsmykla commented on github-event/handler.go:
```diff
@@ -238,3 +268,56 @@ func forward(req []byte, function string, headers map[string]string) (string, in
+
+		for i, c := range customers {
+			customers[i] = strings.ToLower(c)
+			customers[i] = strings.TrimSuffix(c, "\r")
```
There is something wrong, you are assigning value and then immediately overriding it

martindekov replied:
Yep this was irrational 😄

martindekov commented:
I amended some fixes, I haven't looked into the function I just moved the logic out. Also added more detailed reports since those were only from the requests. Thanks for the feedback @bartsmykla really helped out
